ui/src/controller/vsync_controller.ts

```diff
@@ -0,0 +1,121 @@
+// Copyright (C) 2022 The Android Open Source Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {assertExists} from '../base/logging';
+import {Actions} from '../common/actions';
+import {Engine} from '../common/engine';
+import {LONG, NUM} from '../common/query_result';
+import {Span} from '../common/time';
+import {globals} from '../frontend/globals';
+import {Controller} from './controller';
+
+export interface VsyncControllerArgs {
+  engine: Engine;
+}
+
+export class VsyncController extends Controller<'init'|'ready'> {
+  private engine: Engine;
+  private trackId?: number;
+  private timeSpan?: Span<bigint>;
+
+  constructor(args: VsyncControllerArgs) {
+    super('init');
+
+    this.engine = args.engine;
+  }
+
+  onDestroy() {
+    globals.dispatch(Actions.setVsyncData({}));
+  }
+
+  run() {
+    switch (this.state) {
+      case 'init':
+        this.getVsyncTrack().then((trackId) => this.trackId = trackId);
+        this.setState('ready');
+        break;
+      case 'ready':
+        if (this.trackId === undefined) {
+          // Still waiting for the track ID
+          return;
+        }
+        const visibleTimeSpan = globals.stateVisibleTime();
+        if (this.timeSpan === undefined ||
+            !this.timeSpan.equals(visibleTimeSpan)) {
+          this.timeSpan = visibleTimeSpan;
+          this.getVsyncData(this.trackId, this.timeSpan);
+        }
+        break;
+      default:
+        throw new Error(`Unexpected state ${this.state}`);
+    }
+  }
+
+  async getVsyncTrack(): Promise<number|undefined> {
+    // Determine the track ID of the SurfaceFlinger VSYNC-app counter, if
+    // it exists.
+    const result = await this.engine.query(`
+      select process_counter_track.id as trackId
+      from process
+        join process_counter_track using (upid)
+      where process.name like '%/surfaceflinger' and
+        process_counter_track.name='VSYNC-app'
+      limit 1;
+    `);
+    if (result.numRows() < 1) {
+      return undefined;
+    }
+
+    const trackId = result.firstRow({'trackId': NUM}).trackId;
+    return trackId;
+  }
+
+  async getVsyncData(trackId: number, timeSpan: Span<bigint>) {
+    // Get at least two changes of the counter in the currently
```

Comment:
Maybe I'm reading the implementation wrong, but it seems this could result with a single change instead of at least two.

Also, `get` in the name seems a little bit misleading since nothing is returned.